metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java

```diff
@@ -42,326 +36,415 @@
-    return DereferenceUtils.extractReferencesForDereferencing(rdf);
-  }
-
-  private DereferencedEntity dereferenceOwnEntities(Set<ReferenceTerm> resourceIds) {
-    HashSet<Report> reports = new HashSet<>();
-    if (entityResolver == null) {
-      return new DereferencedEntity(Collections.emptyList(), reports);
+
+    @Override
+    public List<DereferencedEntities> dereferenceEntities(Map<Class<? extends AboutType>, Set<String>> resourceIds) {
+
+        // Sanity check.
+        if (resourceIds.isEmpty()) {
+            return List.of(new DereferencedEntities(Collections.emptyMap(), new HashSet<>()));
+        }
+
+        // First try to get them from our own entity collection database.
+        HashMap<Class<? extends AboutType>, Set<ReferenceTerm>> mappedReferenceTerms = new HashMap<>();
+        HashMap<Class<? extends AboutType>, DereferencedEntities> dereferencedEntities = new HashMap<>();
+        resourceIds.forEach((key, value) -> {
+            HashSet<Report> reports = new HashSet<>();
+            Set<ReferenceTerm> referenceTermSet = setUpReferenceTermSet(value, reports);
+            mappedReferenceTerms.put(key, referenceTermSet);
+            final DereferencedEntities dereferencedOwnEntities = dereferenceOwnEntities(referenceTermSet, reports, key);
+            dereferencedEntities.put(key, dereferencedOwnEntities);
+        });
+
+        final Set<String> foundOwnEntityIds = dereferencedEntities
+                .values().stream()
+                .map(DereferencedEntities::getReferenceTermListMap)
+                .map(Map::values).flatMap(Collection::stream).flatMap(Collection::stream)
+                .map(EnrichmentBase::getAbout)
+                .collect(Collectors.toSet());
+
+        //TODO Sorted because unit tests were failing due to "incorrect" order
+        //TODO The order should not matter, the unit tests need fixing
+        //TODO There's already ticket MET-5056 to handle it
+        List<Map.Entry<Class<? extends AboutType>, Set<ReferenceTerm>>> sortedList = mappedReferenceTerms.entrySet()
+                .stream().sorted(Comparator.comparing(entry -> entry.getKey().getName()))
+                .collect(Collectors.toList());
+
+        // For the remaining ones, get them from the dereference service.
+        for (Map.Entry<Class<? extends AboutType>, Set<ReferenceTerm>> entry : sortedList) {
+            DereferencedEntities dereferencedResultEntities = new DereferencedEntities(new HashMap<>(), new HashSet<>());
+
+            if(entry.getKey().equals(Aggregation.class)){
+                Optional<DereferencedEntities> aggregationDereferencingResult =
+                        dereferenceAggregationReferenceTerms(entry.getValue(), foundOwnEntityIds, entry.getKey());
+                if(aggregationDereferencingResult.isPresent()){
+                    updateDereferencedEntitiesMap(dereferencedEntities, entry.getKey(), aggregationDereferencingResult.get());
```

Comment:
There could be some re-usability yes, but when dereferencing with Uri, one of the input parameters is `Set<ReferenceTerm>` (that's why later I did a `containsAll` incorrectly). I was trying to avoid making too many calls with a set of one element